index.js

```diff
@@ -502,6 +506,452 @@
+      {f: [Z.Applicative], t: [Z.Traversable]},
+      [TypeRep($.TypeVariable('f')), t(f(b)), f(t(b))],
+      Z.sequence);
+
+  //# ap :: Apply f => f (a -> b) -> f a -> f b
+  //.
+  //. Curried version of [`Z.ap`][].
+  //.
+  //. ```javascript
+  //. > S.ap([Math.sqrt, x => x * x], [1, 4, 9, 16, 25])
+  //. [1, 2, 3, 4, 5, 1, 16, 81, 256, 625]
+  //.
+  //. > S.ap(s => n => s.slice(0, n), s => Math.ceil(s.length / 2))('Haskell')
+  //. 'Hask'
+  //.
+  //. > S.ap(S.Just(Math.sqrt), S.Just(64))
+  //. Just(8)
+  //. ```
+  S.ap =
+  def('ap',
+      {f: [Z.Apply]},
+      [f(Fn(a, b)), f(a), f(b)],
+      Z.ap);
+
+  //# apFirst :: Apply f => f a -> f b -> f a
+  //.
+  //. Curried version of [`Z.apFirst`][]. Combines two effectful actions,
+  //. keeping only the result of the first. Equivalent to Haskell's `(<*)`
+  //. function.
+  //.
+  //. See also [`apSecond`](#apSecond).
+  //.
+  //. ```javascript
+  //. > S.apFirst([1, 2], [3, 4])
+  //. [1, 1, 2, 2]
+  //.
+  //. > S.apFirst(S.Just(1), S.Just(2))
+  //. Just(1)
+  //. ```
+  S.apFirst = def('apFirst', {f: [Z.Apply]}, [f(a), f(b), f(a)], Z.apFirst);
+
+  //# apSecond :: Apply f => f a -> f b -> f b
+  //.
+  //. Curried version of [`Z.apSecond`][]. Combines two effectful actions,
+  //. keeping only the result of the second. Equivalent to Haskell's `(*>)`
+  //. function.
+  //.
+  //. See also [`apFirst`](#apFirst).
+  //.
+  //. ```javascript
+  //. > S.apSecond([1, 2], [3, 4])
+  //. [3, 4, 3, 4]
+  //.
+  //. > S.apSecond(S.Just(1), S.Just(2))
+  //. Just(2)
+  //. ```
+  S.apSecond = def('apSecond', {f: [Z.Apply]}, [f(a), f(b), f(b)], Z.apSecond);
+
+  //# of :: Applicative f => TypeRep f -> a -> f a
+  //.
+  //. Curried version of [`Z.of`][].
+  //.
+  //. ```javascript
+  //. > S.of(Array, 42)
+  //. [42]
+  //.
+  //. > S.of(Function, 42)(null)
+  //. 42
+  //.
+  //. > S.of(S.Maybe, 42)
+  //. Just(42)
+  //.
+  //. > S.of(S.Either, 42)
+  //. Right(42)
+  //. ```
+  S.of =
+  def('of',
+      {f: [Z.Applicative]},
+      [TypeRep($.TypeVariable('f')), a, f(a)],
+      Z.of);
+
+  //# chain :: Chain m => (a -> m b) -> m a -> m b
+  //.
+  //. Curried version of [`Z.chain`][].
+  //.
+  //. ```javascript
+  //. > S.chain(x => [x, x], [1, 2, 3])
+  //. [1, 1, 2, 2, 3, 3]
+  //.
+  //. > S.chain(n => s => s.slice(0, n), s => Math.ceil(s.length / 2))('slice')
+  //. 'sli'
+  //.
+  //. > S.chain(S.parseInt(10), S.Just('123'))
+  //. Just(123)
+  //.
+  //. > S.chain(S.parseInt(10), S.Just('XXX'))
+  //. Nothing
+  //. ```
+  S.chain = def('chain', {m: [Z.Chain]}, [Fn(a, m(b)), m(a), m(b)], Z.chain);
+
+  //# join :: Chain m => m (m a) -> m a
+  //.
+  //. [Type-safe][sanctuary-def] version of [`Z.join`][].
+  //. Removes one level of nesting from a nested monadic structure.
+  //.
+  //. ```javascript
+  //. > S.join([[1], [2], [3]])
+  //. [1, 2, 3]
+  //.
+  //. > S.join([[[1, 2, 3]]])
+  //. [[1, 2, 3]]
+  //.
+  //. > S.join(S.concat)('abc')
+  //. 'abcabc'
```

Comment:
Yesterday when I asked about `\x -> f x x` @brendanhay told me that `join` is the W combinator! Very cool! See [Stack Overflow][1] if you're interested in the substitutions which show that `S.join(f)` is equivalent to `x => f(x)(x)`.


[1]: http://stackoverflow.com/q/38401606/312785